src/totalpower.cpp

```diff
@@ -0,0 +1,350 @@
+   for(auto it = cbegin();it!=cend();it++){
+     printf("%.3f ",it->median);
+   }
+   printf("\n");
+   if( size() >= m_MaxElements ){
+      pop_front(); // remove element in front
+   }
+   push_back( tmp );
+
+   assert( size() <= m_MaxElements );
+
+   // Sort median and rmsiqr values to get median of medians and median of RMSQRSs
+   // TODO: ? probably there could be some more efficient and clever way of "insertion" like into a binary tree.
+   //       or implement fit or Savitzky-Goley or perfect smoothing (see Adrian S.)
+   // once it works, I can implement improvements :
+   //    - keep array of values in double* array and list of indexes in queue
+   //    - add new element : remove index from the front, push to back of queue of
+   //                        indexes , replaces in array with new element,
+   //                        move the element like in bubble sort to find its place , update
+   //                        indexes in the queue would be the most difficult part in this way
+   // sort Median values :
+   sorting_table.clear();
+   for(auto it = cbegin();it!=cend();it++){
+      sorting_table.push_back(it->median);
+   }
+   ::sort(sorting_table.begin(),sorting_table.end());
+   
+   printf("DEBUG_MEDIAN_OF_MEDIANS:\n");
+   for(int k=0;k<sorting_table.size();k++){
+      printf("%.3f ",sorting_table[k]);
+   }
+   m_MedianOfMedians = sorting_table[int(sorting_table.size()/2)];
+   printf("\n -> median_of_medians = %.3f\n",m_MedianOfMedians);
+   
+   sorting_table.clear();
+   for(auto it = cbegin();it!=cend();it++){
+      sorting_table.push_back(it->rmsiqr);
+   }
+   ::sort(sorting_table.begin(),sorting_table.end());
+
+   
+   m_MedianOfRmsIQRs = sorting_table[int(sorting_table.size()/2)];     
+}
+
+TotalPowerRec::TotalPowerRec()
+: time_index(-1), total_power(0.00)
+{
+}
+
+TotalPower::TotalPower( double time_resolution_ms, int history_size, double start_freq_mhz, double end_freq_mhz )
+: m_Median(0.00), m_RMSIQR(0.00), m_StartFreqMHz(start_freq_mhz), m_EndFreqMHz(end_freq_mhz), m_TimeResolutionMS(time_resolution_ms), m_MedianIQRHistory(history_size)
+{
+   printf("DEBUG : created TotalPower object in %.4f - %.4f MHz range, time resolution = %.4f [ms], will keep %d samples history\n",start_freq_mhz,end_freq_mhz,m_TimeResolutionMS,history_size);
+}
+
+TotalPower::~TotalPower()
+{
+}
+
+
+double TotalPower::GetMedianOfMedians()
+{
+   return m_MedianIQRHistory.m_MedianOfMedians;
+}
+
+double TotalPower::GetMedianOfRMSIQRs()
+{
+   return m_MedianIQRHistory.m_MedianOfRmsIQRs;
+}
+
+void TotalPower::ResetFile()
+{
+   // just clean up file 
+   FILE* out_f = fopen(filename.c_str(),"w");
+   // rec.time_index,rec.total_power,GetMedianOfMedians(),GetMedianOfRMSIQRs(),m_Median);
+   fprintf(out_f,"# TIMEINDEX TOTAL_POWER MedianOfMedians MedianOfRMSIQR MEDIAN\n");
+   fclose(out_f);
+}
+
+
+void TotalPower::calc( DynamicSpectrum& dynaspec, bool do_dump, bool use_rms, int offset, int ntimes, long int start_time )
```

Comment:
The FITS class and DynamicSpectrum class could be abstracted away if you passed the `float *data`, `size_t n_timesteps` and `size_t n_channels` as parameters of the method, I think. So you write the logic only once.

Something like:

```
void actual_calc(float *ds, size_t n_timesteps, size_t n_channels, ...) {...}

void calc(DynamicSpectrum& dynaspec, ...) {
    actual_calc(dynaspec.data(), dynastic.n_timesteps(), ...)
}

void calc(CBgFits& dynaspec, ...) {
    actual_calc(dynaspec.data(), dynastic.x_size(), ...)
}
```